src/App.tsx

```diff
@@ -1,14 +1,89 @@
-/* eslint-disable max-len */
-import React from 'react';
+import React, {
+  useEffect,
+  useState,
+  useCallback,
+  useMemo,
+} from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { ErrorMessage } from './components/ErrorMessage';
+
+import { Todo } from './types/Todo';
+import { TodoWithUser } from './types/TodoWithUser';
+import { TodoCompletionFilter } from './types/TodoCompletionFilter';
+
+import { getTodos, getUser } from './api';
+
+import { filterTodos } from './helpers';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isTodosLoading, setIsTodosLoading] = useState(false);
+  const [hasTotosLoadingError, setHasTodosLoadingError] = useState(false);
+  const [isTodosLoadInitialized, setIsTodosLoadInitialized] = useState(false);
+
+  const [selectedTodoWithUser, setSelectedTodoWithUser]
+    = useState<null | TodoWithUser>(null);
+  const [isTodoWithUserLoading, setIsTodoWithUserLoading]
+    = useState(false);
+  const [hasTodoWithUserLoadingError, setHasTodoWithUserLoadingError]
+    = useState(false);
+  const [isTodosWithUserLoadInitialized, setIsTodosWithUserLoadInitialized]
+    = useState(false);
+
+  const [todoCompletionFilterOption, setTodoCompletionFilterOption]
+    = useState(TodoCompletionFilter.All);
+  const [searchQuery, setSearchQuery] = useState('');
+
+  useEffect(() => {
+    setIsTodosLoading(true);
+    setIsTodosLoadInitialized(true);
+
+    getTodos()
+      .then(setTodos)
+      .catch(() => setHasTodosLoadingError(true))
+      .finally(() => setIsTodosLoading(false));
+  }, []);
+
+  const handleTodoWithUserSelect = useCallback(
+    (newSelectedTodo: Todo): void => {
+      setHasTodoWithUserLoadingError(false);
+      setIsTodoWithUserLoading(true);
+      setIsTodosWithUserLoadInitialized(true);
+      setSelectedTodoWithUser({
+        ...newSelectedTodo,
+        user: null,
+      });
+
+      getUser(newSelectedTodo.userId)
+        .then(user => setSelectedTodoWithUser({
+          ...newSelectedTodo,
+          user: { ...user },
+        }))
+        .catch(() => setHasTodoWithUserLoadingError(true))
+        .finally(() => setIsTodoWithUserLoading(false));
+    },
+    [],
+  );
```

Comment:
Sorry, but it's a bit hard to read. Try in future to make more simple and useful code

Reply:
+ try to keep the app more lightweight. Use helpers in different folders or files.

Reply:
get used to